### PR TITLE
perf: adopt SWAR-accelerated utils helpers in hot-path scanners

### PR DESCRIPTION
**Pull request by gaby**
# Description

gofiber/utils v2.1.2 added SWAR (SIMD-within-a-register) accelerated search helpers and a public `swar` primitives package. This PR adopts them across fiber's hot-path scanners — replacing hand-rolled per-byte loops with word-at-a-time equivalents. Every change was gated on `benchstat` (n=10); candidates that regressed were reverted and documented rather than shipped. Behavior is preserved exactly at every site except two deliberate correctness fixes called out below, and the final state was verified by multi-million-input differential fuzzing against the original implementations across three review rounds. A full merge-base-vs-branch benchmark matrix is posted in the PR comments.

**Header parsing (per request):**
- `forEachMediaRange`: the quoted Accept-header branch skip-scans with `utils.IndexAny3(',', '"', '\\')` instead of visiting every byte
- `isEtagStale`: If-None-Match splits on `utils.IndexAny2('"', ',')` hops
- `quoteRawString`: escape detection is a single `swar` scan (`indexQuoteEscape`) whose mask is exactly the escape set — RFC 9110 non-quotable bytes plus HTAB, which the RFC permits as qdtext but this function has always percent-encoded — and the slow path bulk-copies the clean prefix
- duplicate hand-rolled `isASCII` loops (core + hostauthorization) → `utils.IsASCII`

**Routing hot path (first direct use of the raw `swar` primitives):**
- `configDependentPaths` fuses the detection-path copy with its ASCII case fold into a single pass (`appendLowerASCII`, `swar.Load8/ToLowerWord/Store8`, amortized growth via `slices.Grow`), removing one full write pass per request in the default `CaseSensitive: false` config
- the `alpha` route constraint validates eight bytes per iteration with two `swar.MatchRangeMask` lane checks, falling back to the existing `unicode.IsLetter` rune loop on the first non-ASCII byte (handoff always lands on a rune boundary)

**Logging:**
- `log` context writer: control-byte detection in `writeSanitized`/`writeSanitizedString` (per context-tag value on every log line) is now a `swar` word scan returning the first offending index, and the scrub copies the clean prefix wholesale via a shared generic `scrubControls`

**Middleware and binder:**
- basicauth: `containsInvalidHeaderChars` swaps `strings.IndexFunc` for a `swar` word scan over the raw Authorization header (valid lane = visible ASCII or HTAB); `containsCTL` scans ASCII spans word-at-a-time and defers to the `unicode.IsControl` rune loop at the first byte ≥ 0x80, keeping multi-byte C1 detection intact; the post-scheme whitespace check becomes `utils.IndexAny2(' ', '\t')` (justified by the preceding header validation, cross-referenced at both sites). All run before the credential compare, so constant-time properties are unaffected
- cache: `hasDirective` drops the per-offset `EqualFold` loop for a first-byte fold filter + `EqualFold` verify — the same shape that won for `isNoCache` (an interim `IndexFold` version measured slower and was replaced)
- requestid: `isValidRequestID` checks the visible-ASCII range eight bytes at a time with `swar.MatchRangeMask`
- proxy: `resolveRedirect` rejects Location control bytes via a `swar` scan (`containsCTLOrDEL`; bytes ≥ 0x80 stay allowed, unchanged)
- static: `sanitizePath` rejects backslashes and NUL in one `utils.IndexAny2` pass instead of two `IndexByte` scans
- sse `sanitizeField`, cache `escapeKeyDelimiters`, binder `FilterFlags`: `strings.ContainsAny`/`IndexAny` (per-byte asciiSet walks) → `utils.IndexAny2/3`

**Two deliberate behavior fixes** (found while reviewing the touched code, each with regression tests):
- `forEachMediaRange` now treats a quoted-pair as escaping exactly one byte per RFC 9110 §5.6.4. Previously the escape flag was never reset, so a genuine closing quote after `\"` was itself treated as escaped and the rest of the Accept header collapsed into a single media range (e.g. `text/html;p="a\"b", text/plain` failed to match a `text/plain` offer). The new behavior also agrees with fasthttp's header-param parsing and fiber's own `Charset` parser.
- cache `hasDirective` returns false for an empty directive instead of relying on callers never passing one (both call sites pass constants; the guard makes the helper total).

**Measured and rejected** (kept out deliberately): rewriting `isNoCache` with `utils.IndexFold` regressed ~2x on realistic short Cache-Control values, and a `swar` rewrite of extractors' `isValidToken68` benchmarked +16% against the compiler's optimized scalar switch; both originals stay, with the results recorded in commit messages and code comments. `Charset()` and `extractIPsFromHeader` were examined and skipped — their state machines react to every ordinary byte, so a skip-scan is not behavior-preserving. The remaining router scans (`findParamLen*`, `pathSlashCount`, route compares) already lower to runtime SIMD (`IndexByte`/`Index`/`Count`/`memequal`) and were left untouched, as were timing-sensitive credential/token comparisons (`crypto/subtle` paths) across all middleware.

## Changes introduced

- [x] Benchmarks: full merge-base-vs-branch matrix with identical harnesses on both sides posted in the PR comments (`benchstat`, n=10). Micro highlights: cache `hasDirective` −83%, basicauth `containsInvalidHeaderChars` −81%, basicauth `containsCTL` −74%, sse `sanitizeField` −70%, requestid `isValidRequestID` −65%, log `writeSanitizedString` −65%. End-to-end: `ConstraintExecution/alpha` −16%, `Ctx_Attachment` −11.5%, `Middleware_BasicAuth` −9.4%, `Router_Handler` −7.7%, `Accepts*` −2…6%, allocations unchanged everywhere (0 B/op delta on all touched paths)
- [x] Changelog/What's New: performance — hot-path header parsing, routing path normalization, log sanitization, and header validation in the basicauth/cache/requestid/proxy/static middleware now use SWAR-accelerated scanning from gofiber/utils v2.1.2; fix — quoted-pairs in Accept-header media ranges now follow RFC 9110 §5.6.4

(Documentation, migration guide, Express alignment, examples: not applicable — no public API changes.)

## Type of change

- [x] Performance improvement (non-breaking change which improves efficiency)
- [x] Code consistency (non-breaking change which improves code reliability and robustness)

## Checklist

- [x] Conducted a self-review of the code and provided comments for complex or critical parts (three adversarial review rounds with differential fuzzing; final round clean).
- [x] Added or updated unit tests to validate the effectiveness of the changes or new features (HTAB escaping lock-in, quoted-pair regression case, `appendLowerASCII` word/tail/non-ASCII/buffer-reuse cases, alpha constraint word-boundary + Unicode fallback cases, requestid/basicauth word-boundary cases, `containsCTL` C1-control cases, empty-directive cases, `resolveConstraintName` Unicode-folding pin).
- [x] Ensured that new and existing unit tests pass locally with the changes (full suite with `-race -shuffle=on`, `golangci-lint` clean).
- [x] Verified that any new dependencies are essential and have been agreed upon by the maintainers/community (none added — uses the already-pinned gofiber/utils v2.1.2).
- [x] Aimed for optimal performance with minimal allocations in the new code.
- [x] Provided benchmarks for the new code to analyze and improve upon (added `Benchmark_isValidRequestID`, `Benchmark_containsInvalidHeaderChars`, `Benchmark_containsCTL`, `Benchmark_hasDirective`, `Benchmark_writeSanitizedString`, `Benchmark_isValidToken68`, `Benchmark_sanitizeField`; existing router/ctx/helpers benchmarks cover the rest).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01XQHTh1nZDARUPzNDBsba6h

### CODE REVIEW

File: middleware/proxy/proxy.go [hunk 1]
```diff
@@ -10,6 +10,7 @@ import (
 	"time"
 
 	"github.com/gofiber/utils/v2"
+	"github.com/gofiber/utils/v2/swar"
```
[the remark this answers is not in the document]
Comment (gaby) [reply middleware/proxy/proxy.go:13]: The base import was never removed — `middleware/proxy/proxy.go` imports both `github.com/gofiber/utils/v2` and `github.com/gofiber/utils/v2/swar` (lines 12-13), and `utils.EqualFold` at lines 451/531 compiles fine. The lint and ubuntu unit jobs passed on this commit. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by ReneWerner87** [comment]
@gaby I saw a few benchmarks above
could you organize them even better and make sure to cover all the parts that were adjusted (if possible)?
Also, please post the parts that stayed the same but were actually adjusted.

**Comment by gaby** [comment]
@ReneWerner87 Here is the complete, organized benchmark matrix covering every adjusted part, including the ones that are performance-neutral.

**Methodology:** baseline = merge-base `6a52c1d` in a pristine worktree, head = branch tip. The new micro-benchmarks were planted into the baseline tree so old and new implementations are measured by byte-identical harnesses. `benchstat`, n=10 per side, sequential runs, linux/amd64 (Xeon 2.1 GHz). `~` = no statistically significant difference.

### 1. Rewritten functions (micro-benchmarks, old implementation vs new)

| Function (file) | old | new | Δ |
|---|---|---|---|
| `hasDirective` (middleware/cache) | 438.8n | 72.6n | **−83.5%** |
| `containsInvalidHeaderChars` (middleware/basicauth) | 141.8n | 26.3n | **−81.5%** |
| `containsCTL` (middleware/basicauth) | 81.3n | 21.1n | **−74.0%** |
| `sanitizeField` (middleware/sse) | 99.8n | 29.5n | **−70.4%** |
| `isValidRequestID` (middleware/requestid) | 43.9n | 15.3n | **−65.2%** |
| `writeSanitizedString` (log) | 166.6n | 59.1n | **−64.5%** |

### 2. End-to-end request paths (existing benchmarks, merge-base vs branch)

| Benchmark | Δ | driven by |
|---|---|---|
| `ConstraintExecution/alpha` | **−16.0%** | alpha constraint SWAR fast path |
| `Ctx_Attachment` | **−11.5%** | `quoteRawString` + `utils.IsASCII` |
| `Middleware_BasicAuth` (sha256 creds, full handler) | **−9.4%** | the three basicauth scans |
| `Utils_GetOffer/6_offers` | −9.1% | `forEachMediaRange` skip-scan |
| `Router_Handler` | −7.7% | fused lowercase-copy in path normalization |
| `Utils_GetOffer/web_browser` | −6.1% | `forEachMediaRange` |
| `Ctx_Accepts/{json,xml}` | −6.4% | `forEachMediaRange` |
| `Ctx_AcceptsEncodings` | −6.1% | `forEachMediaRange` |
| `Ctx_AcquireReleaseFlow/withRequestCtx` | −4.6% | fused lowercase-copy |
| allocations | 0 B/op change on every touched path | — |

### 3. Adjusted but performance-neutral (measured, as expected)

| Benchmark | Δ | why neutral |
|---|---|---|
| `Ctx_Fresh_StaleEtag` / `Etag` middleware | ~ | `isEtagStale` splits on `IndexAny2` hops; If-None-Match values are short, gain is within noise — kept for the simpler loop |
| `Ctx_Fresh_WithNoCache` / `Utils_IsNoCache` | ~ | `isNoCache` deliberately unchanged (see §4) |
| `Router_Handler_CaseSensitive` | ~ | exercises the branch that does **not** use the fused lowercase-copy — confirms no penalty for that config |
| `Router_Next_Default`, `Route_Match` | ~ | routing match loops untouched (already stdlib SIMD); confirms no collateral cost |
| `Ctx_Links` | +6% (~4ns) | `quoteRawString` call overhead on sub-word values ("next"/"last"); this matrix initially showed +18% from a two-pass fast path — fixed in 05b5a3a with a single fused scan, and the same function gives Attachment its −11.5% |
| `GetOffer/*parameters*` | +2…5% | parameter parsing (`forEachParameter`) is untouched by this PR; mixed ± across untouched subs on this runner points to code-layout noise, not a regression in changed code |

### 4. Attempted, measured slower, reverted (originals kept)

| Function | attempted | result | state |
|---|---|---|---|
| `isNoCache` (helpers.go) | `utils.IndexFold` skip-scan | +93% (46.8n → 90.6n) | original first-byte-filter kept; rationale in revert commit |
| `isValidToken68` (extractors) | full `swar` character-class scan | +16% | original scalar kept; NOTE comment in code records the measurement (matrix confirms parity: −1.8% ≈ noise) |
| `hasDirective` interim | `utils.IndexFold` skip-scan | slower than the first-byte-filter shape (86.9n vs 74.0n) | final version uses the same shape that won for `isNoCache` |

### 5. Adjusted without a dedicated benchmark (and why)

- `containsCTLOrDEL` (proxy `resolveRedirect`): internal to the redirect-following flow; same scanner shape as the six benchmarked ones, predicate covered by unit tests. Happy to add a micro-benchmark if you want it gated too.
- `sanitizePath` (static), `FilterFlags` (binder), `escapeKeyDelimiters` (cache), `toPunycode` ASCII check (hostauthorization): single-call swaps to `utils.IndexAny2/3`/`utils.IsASCII`, whose primitives are benchmarked upstream in gofiber/utils; behavior covered by existing package tests (static path-traversal suite, binder/cache suites).
- `appendLowerASCII` (path normalization): no standalone benchmark, but it is the driver behind the `Router_Handler` −7.7% and `AcquireReleaseFlow` −4.6% rows above and has a dedicated unit test for word/tail/non-ASCII/buffer-reuse cases.

Raw benchstat output available on request.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by ReneWerner87 — APPROVED** [review]
